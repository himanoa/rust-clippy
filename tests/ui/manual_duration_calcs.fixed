// run-rustfix
#![allow(dead_code)]
#![warn(clippy::manual_duration_calcs)]

use std::time::Duration;

fn main() {
    let dur = Duration::new(5, 0);

    let bad_millis_1 = dur.subsec_millis();
    let bad_millis_2 = dur.subsec_millis();
    let good_millis = dur.subsec_millis();
    assert_eq!(bad_millis_1, good_millis);
    assert_eq!(bad_millis_2, good_millis);

    let bad_micros = dur.subsec_micros();
    let good_micros = dur.subsec_micros();
    assert_eq!(bad_micros, good_micros);

    // Handle refs
    let _ = (&dur).subsec_micros();

    // Handle constants
    const NANOS_IN_MICRO: u32 = 1_000;
    let _ = dur.subsec_micros();

    // Other literals aren't linted
    let _ = dur.subsec_nanos() / 699;

    // Manual implementation
    let _nanos = dur.as_nanos() as u64;
    let _milis = dur.as_millis() as u64;
    let _secs_f64_1 = dur.as_secs_f64();
    let _secs_f64_2 = dur.as_secs_f32();
    let _secs_f64_3 = dur.as_secs_f64();
    let _secs_f64_4 = dur.as_secs_f64();

    let _secs_f32_2 = dur.as_secs_f32();
    let _secs_f32_3 = dur.as_secs_f32();
}
