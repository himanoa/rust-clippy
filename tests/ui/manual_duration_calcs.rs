// run-rustfix
#![allow(dead_code)]
#![warn(clippy::manual_duration_calcs)]

use std::time::Duration;

fn main() {
    let dur = Duration::new(5, 0);

    let bad_millis_1 = dur.subsec_micros() / 1_000;
    let bad_millis_2 = dur.subsec_nanos() / 1_000_000;
    let good_millis = dur.subsec_millis();
    assert_eq!(bad_millis_1, good_millis);
    assert_eq!(bad_millis_2, good_millis);

    let bad_micros = dur.subsec_nanos() / 1_000;
    let good_micros = dur.subsec_micros();
    assert_eq!(bad_micros, good_micros);

    // Handle refs
    let _ = (&dur).subsec_nanos() / 1_000;

    // Handle constants
    const NANOS_IN_MICRO: u32 = 1_000;
    let _ = dur.subsec_nanos() / NANOS_IN_MICRO;

    // Other literals aren't linted
    let _ = dur.subsec_nanos() / 699;

    // Manual implementation
    let _nanos = diff.as_secs() * 1_000_000_000 + diff.subsec_nanos() as u64;
    let _milis = diff.as_secs() * 1_000 + diff.subsec_milis() as u64;
    let _secs_f64_1 = (diff.as_secs() as f64 * 1_000.0 + diff.subsec_milis() as f64) / 100.0;
    let _secs_f64_2 = diff.as_secs() as f64 + diff.subsec_milis() as f64 / 1_000.0;
    let _secs_f64_3 = diff.as_secs() as f64 + diff.subsec_nanos() as f64 / 1_000_000_000.0;
}
