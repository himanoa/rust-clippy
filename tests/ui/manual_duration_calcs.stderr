error: calling `subsec_millis()` is more concise than this calculation
  --> $DIR/manual_duration_calcs.rs:10:24
   |
LL |     let bad_millis_1 = dur.subsec_micros() / 1_000;
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.subsec_millis()`
   |
   = note: `-D clippy::manual-duration-calcs` implied by `-D warnings`

error: calling `subsec_millis()` is more concise than this calculation
  --> $DIR/manual_duration_calcs.rs:11:24
   |
LL |     let bad_millis_2 = dur.subsec_nanos() / 1_000_000;
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.subsec_millis()`

error: calling `subsec_micros()` is more concise than this calculation
  --> $DIR/manual_duration_calcs.rs:16:22
   |
LL |     let bad_micros = dur.subsec_nanos() / 1_000;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.subsec_micros()`

error: calling `subsec_micros()` is more concise than this calculation
  --> $DIR/manual_duration_calcs.rs:21:13
   |
LL |     let _ = (&dur).subsec_nanos() / 1_000;
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `(&dur).subsec_micros()`

error: calling `subsec_micros()` is more concise than this calculation
  --> $DIR/manual_duration_calcs.rs:25:13
   |
LL |     let _ = dur.subsec_nanos() / NANOS_IN_MICRO;
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.subsec_micros()`

error: no manual re-implementationa of the as_nanos
  --> $DIR/manual_duration_calcs.rs:31:18
   |
LL |     let _nanos = dur.as_secs() * 1_000_000_000 + dur.subsec_nanos() as u64;
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.as_nanos() as u64`

error: no manual re-implementationa of the as_millis
  --> $DIR/manual_duration_calcs.rs:32:18
   |
LL |     let _milis = dur.as_secs() * 1_000 + dur.subsec_millis() as u64;
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.as_millis() as u64`

error: no manual re-implementationa of the as_secs_f64
  --> $DIR/manual_duration_calcs.rs:33:23
   |
LL |     let _secs_f64_1 = (dur.as_secs() as f64 * 1_000.0 + dur.subsec_millis() as f64) / 1000.0;
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.as_secs_f64()`

error: no manual re-implementationa of the as_secs_f64
  --> $DIR/manual_duration_calcs.rs:34:23
   |
LL |     let _secs_f64_2 = dur.as_secs() as f64 + dur.subsec_millis() as f64 / 1_000.0;
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.as_secs_f64()`

error: no manual re-implementationa of the as_secs_f64
  --> $DIR/manual_duration_calcs.rs:35:23
   |
LL |     let _secs_f64_3 = dur.as_secs() as f64 + dur.subsec_nanos() as f64 / 1_000_000_000.0;
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.as_secs_f64()`

error: no manual re-implementationa of the as_secs_f32
  --> $DIR/manual_duration_calcs.rs:37:23
   |
LL |     let _secs_f32_2 = dur.as_secs() as f32 + dur.subsec_millis() as f32 / 1_000.0;
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.as_secs_f32()`

error: no manual re-implementationa of the as_secs_f32
  --> $DIR/manual_duration_calcs.rs:38:23
   |
LL |     let _secs_f32_3 = dur.as_secs() as f32 + dur.subsec_nanos() as f32 / 1_000_000_000.0;
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.as_secs_f32()`

error: aborting due to 12 previous errors

